### DISCUSSION TITLE
Merge option for generator output

### DISCUSSION THREAD

**Discussion by jondavid-black** [DevOps-MBSE/AaC#699] [Ideas]
I don't know if this would be good or not, but it seems like an ability to merge generated content into existing code would be handy.

I spent a little time experimenting with this and discovered it is quite complex.  I used GitHub Copilot to generate some code for me, but didn't get to anything that worked properly.  I also realized that different merge logic may be needed for different AaC roots.

Here's an example of the AI prompt and output:

### Prompt: 

Let's take a fresh start and see if we can oversome some confusion. I want a merge function that takes in a str called template_output containing generated code and a file name called existing_file with old code. The merge function needs to compare the two and merge them in a similar way to how git will merge code changes. Imports should be the combination of template_output and existing with no duplicates. If there are any module variables defined in the template or existing, the output should be a combination of both without duplicates. Functions should be compared for duplicates. Any function that isn't a duplicate should be output unchanged. Any function that is a duplicate should use the name, args, decorator_list, and returns from the template but use the body from the existing. Be sure to write out the def line for each function with the name, args, returns, and decorator_list. Once merged as described here, write the merged content to the existing_file.

### Output:

Thank you for the clarification. Here's an implementation of the merge_files function that should meet your requirements:

```python
import ast
import difflib

def merge_files(template_output: str, existing_file: str):
    with open(existing_file, 'r') as f:
        existing_contents = f.read()

    if template_output == existing_contents:
        # The template output is identical to the existing file, so do nothing
        return

    # Parse the ASTs of the template output and existing file
    template_ast = ast.parse(template_output)
    existing_ast = ast.parse(existing_contents)

    # Merge the imports
    template_imports = {ast.dump(node) for node in template_ast.body if isinstance(node, ast.Import) or isinstance(node, ast.ImportFrom)}
    existing_imports = {ast.dump(node) for node in existing_ast.body if isinstance(node, ast.Import) or isinstance(node, ast.ImportFrom)}
    new_imports = template_imports - existing_imports
    import_lines = [ast.unparse(node).strip() + '\n' for node in existing_ast.body if isinstance(node, ast.Import) or isinstance(node, ast.ImportFrom) and ast.dump(node) not in new_imports]
    import_lines.extend([ast.unparse(node).strip() + '\n' for node in template_ast.body if isinstance(node, ast.Import) or isinstance(node, ast.ImportFrom) and ast.dump(node) in new_imports])
    import_vars = set()
    for line in import_lines:
        if line.startswith('from '):
            module_name = line.split()[1]
            for node in existing_ast.body:
                if isinstance(node, ast.Assign) and isinstance(node.value, ast.Call) and isinstance(node.value.func, ast.Name) and node.value.func.id == 'ImportFrom' and node.value.args[0].s == module_name:
                    import_vars.update({target.id for target in node.targets})
            for node in template_ast.body:
                if isinstance(node, ast.Assign) and isinstance(node.value, ast.Call) and isinstance(node.value.func, ast.Name) and node.value.func.id == 'ImportFrom' and node.value.args[0].s == module_name:
                    import_vars.update({target.id for target in node.targets})

    # Merge the function definitions
    template_functions = {node.name for node in template_ast.body if isinstance(node, ast.FunctionDef)}
    new_function_defs = {node.name: node for node in template_ast.body if isinstance(node, ast.FunctionDef) and node.name in template_functions}
    function_defs = {node.name: node for node in existing_ast.body if isinstance(node, ast.FunctionDef) and node.name not in template_functions}
    for node in new_function_defs.values():
        function_defs[node.name] = node

    # Write the merged file
    with open(existing_file, 'w') as f:
        f.writelines(import_lines)
        f.write('\n')
        for var in import_vars:
            f.write(f'{var} = {var}\n')
        f.write('\n')
        for node in function_defs.values():
            if node.name in new_function_defs:
                # The function is in the template output, so use the template definition but the existing body
                existing_function_def = function_defs[node.name]
                template_node = new_function_defs[node.name]
                node.name = template_node.name
                node.args = template_node.args
                node.decorator_list = template_node.decorator_list
                node.returns = template_node.returns
                existing_lines = existing_contents.splitlines(keepends=True)
                existing_start = existing_function_def.body[0].lineno - 1
                existing_end = existing_function_def.body[-1].end_lineno - 1
                template_lines = template_output.splitlines(keepends=True)
                template_start = template_node.body[0].lineno - 1
                template_end = template_node.body[-1].end_lineno - 1
                matcher = difflib.SequenceMatcher(None, existing_lines[existing_start:existing_end], template_lines[template_start:template_end])
                operations = matcher.get_opcodes()
                for tag, i1, i2, j1, j2 in operations:
                    if tag == 'equal':
                        # The lines are the same in both files, so just write them to the output file
                        f.writelines(existing_lines[existing_start+i1:existing_start+i2])
                    elif tag == 'replace' or tag == 'insert':
                        # The lines are different in the template output, so write the new lines to the output file
                        f.writelines(template_lines[template_start+j1:template_start+j2])
                    elif tag == 'delete':
                        # The lines are missing in the template output, so skip them
                        pass
            else:
                # The function is not in the template output, so just write it to the output file
                f.write(ast.unparse(node))
```

This implementation should merge the code changes in a similar way to how Git merges code changes. It merges the import statements from both the original and template files without any duplicates. It also includes any module variables from either the template or the original. Finally, it writes the function definitions from the template file, ignoring changes from the original code, and doesn't include any template content in the function body. If a function is a duplicate, it uses the name, args, decorator_list, and returns from the template but uses the body from the existing code.

You can call this function like this:

```python
template_output = 'path/to/template_output.py'
existing_file = 'path/to/existing_file.py'
merge_files(template_output, existing_file)
```

This will merge the changes from the template output file into the existing file and write the result to the same file. If the template output is identical to the existing file, it will do nothing.

### Current Thoughts

The AI output is extremely impressive and certainly provides a path to accomplish a usable merge functionality, but it doesn't result in a valid merged output.  Having spent some time working with the concept, I believe that if we decide to use a merge capability it will need to be handled as a callable passed into the generator that provides the correct behavior for the output type.  For now, I'm just sticking with generator output controls for ```Overwrite``` and ```Skip``` to keep things simple.

Do you think a merge functionality in generation would be valuable?

**Comment by lizzcondrey** [comment]
Conceptually I could see value in covering all CRUD capabilities by including merge. But as you've stated, it is not as straightforward as one would like. It was attempted on a previous project I was on and we never got it completely sorted out. We just had to settle for a modified overwrite that warned the user you were going to lose the previous instance and offer a renaming chance to do manual merging.

**Comment by jondavid-black** [comment]
I beat my head against this for quite a while and decided this isn't feasible.  Instead of trying to merge changes, I opted for generating new content and letting the user worry about migrating changes.  The generator will now create backup files with old code for files marked as OVERWRITE and evaluate code for files marked as SAFE.
